### ISSUE TITLE
LoraModule: rank-dropout shape bug and follow-up robustness notes

### ISSUE THREAD

**Issue by Enferlain** [Enferlain/sd-scripts#93]
## Summary

I found one likely correctness bug in `LoraModule` plus a few follow-up robustness / maintainability concerns worth tracking together.

The main issue is that rank dropout currently assumes the LoRA rank dimension is always axis 1, which is correct for conv outputs but not for general `nn.Linear` outputs with extra dimensions.

## Main issue: rank-dropout assumes the rank dimension is axis 1

In `_apply_rank_dropout`, the mask shape is built like this:

```python
mask_shape = [1] * x.ndim
mask_shape[0] = x.shape[0]
mask_shape[1] = self.lora_dim
```

This works for conv outputs shaped like `[batch, rank, ...]`, but it is not generally correct for `nn.Linear`.

For example, if `Linear` is applied to a tensor shaped like `[batch, seq, hidden]`, then `lora_down(...)` produces `[batch, seq, rank]`, so the rank dimension is the last axis rather than axis 1. In those cases rank dropout will be applied incorrectly.

### Suggested direction

Pick the rank axis by module type:

- `Linear`: use the last dimension as the rank dimension
- `Conv1d` / `Conv2d` / `Conv3d`: use dimension 1 as the rank dimension

Potential sketch:

```python
def _apply_rank_dropout(self, x: Tensor) -> tuple[Tensor, float]:
    if self.rank_dropout <= 0.0 or not self.training:
        return x, self.scale

    mask_shape = [1] * x.ndim
    if self.module_type == "linear":
        rank_dim = x.ndim - 1
    else:
        rank_dim = 1

    mask_shape[0] = x.shape[0]
    mask_shape[rank_dim] = self.lora_dim
    mask = (torch.rand(mask_shape, device=x.device) > self.rank_dropout).to(x.dtype)
    x = x * mask
    scale = self.scale * (1.0 / (1.0 - self.rank_dropout))
    return x, scale
```

## Follow-up notes

These are not necessarily bugs, but they seem worth documenting or revisiting:

### 1. `apply_to()` monkeypatches `forward`

```python
self.org_module[0].forward = self.forward
```

This is a practical pattern and common in LoRA implementations, but it can be fragile around tracing, compilation, wrappers, hooks, or any other code that expects to own `forward`.

### 2. `org_module` is stored in a one-element list

```python
self.org_module = [org_module]
```

The inline comment explains the intent (preserving a mutable reference if runtime code swaps the bound module object), so this is understandable, but it is unusual enough that it should stay clearly documented.

### 3. Explicit fallback dtype behavior is a deliberate semantic choice

`_forward_with_explicit_dtype_fallback()` computes the original branch in the original module dtype and only casts back to `x.dtype` at the end.

That seems reasonable and likely intentional, but it is worth noting that output computation semantics are anchored to the original module dtype rather than the caller input dtype.

### 4. Small readability / cleanup items

- `algo_check()` is fine, but could be made a bit more explicit for readability
- `LoraConfig.alpha` accepting `float | Tensor | None` is flexible, though a bit broader than the normalized runtime representation
- `adapter_target = None` may be dead state if it is unused elsewhere

## Overall assessment

The implementation looks strong overall: the state import/export path is clean, the dtype/autocast handling is more careful than average, and the merge logic looks thoughtful. The one thing that seems important to fix before relying on it broadly is the rank-dropout dimension handling for general `Linear` inputs.
